server/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java

```diff
@@ -140,24 +141,26 @@ static SortedTopDocs sortDocs(
     }
 
     static TopDocs mergeTopDocs(List<TopDocs> results, int topN, int from) {
-        if (results.isEmpty()) {
+        List<TopDocs> topDocsList = results.stream().filter(Objects::nonNull).toList();
```

Comment:
I am still unclear how `null` worked before without the filtering. Stuff breaks pretty badly if you let `null` go through here, but I don't understand how that did not happen before we introduced batched execution, given that we did allow `null`, but I guess we were never calling mergeTopDocs against null results, while we do so now.

Reply:
Exactly we should just not let the `null`s get here was kinda my point initially, I think the way this PR does it now should be fine let me take a look :)

Reply:
We never added `null` to this list because we didn't even queue empty query shard responses for merging. This sort of changed now because suggest+profiling aren't (yet) batched since we have no existing partial merge logic for those.
We can clean this up in a follow-up I had planned which leaves this logic on the transport thread (discussed in the previous PR on this), for now I think doing the filtering down here is fine :)